smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ShapeLink.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.codegen.core;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.NodeMapper;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.ToNode;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+/**
+ * Class that defines a link between the Smithy {@link software.amazon.smithy.model.shapes.Shape} and
+ * the artifact that it produced.
+ */
+public final class ShapeLink implements ToNode, ToSmithyBuilder<ShapeLink> {
+    public static final String TYPE_TEXT = "type";
+    public static final String ID_TEXT = "id";
+    public static final String TAGS_TEXT = "tags";
+    public static final String FILE_TEXT = "file";
+    public static final String LINE_TEXT = "line";
+    public static final String COLUMN_TEXT = "column";
+
+    private String type;
+    private String id;
+    private List<String> tags; //optional
+    private String file; //optional
+    private Integer line; //optional
+    private Integer column; //optional
+
+    private ShapeLink(Builder builder) {
+        type = SmithyBuilder.requiredState(TYPE_TEXT, builder.type);
+        id = SmithyBuilder.requiredState(ID_TEXT, builder.id);
+        tags = ListUtils.copyOf(builder.tags);
+        if (tags.isEmpty()) {
```

Comment:
Lists and Maps are usually left empty instead of set back to null like this one. This prevents needing to route through an `Optional` (like you do below), and requires only a little refactoring in your `toNode()`:

```java
// Store the mostly-complete builder
ObjectNode.Builder nodeBuilder = ObjectNode.objectNodeBuilder()
        .withMember(ID_TEXT, id)
        ...

// Apply the tags member if you have any.
if (!tags.isEmpty()) {
    nodeBuilder.withMember(TAGS_TEXT, Node.fromStrings(tags));
}

// Return the built node
return nodeBuilder.build();
```